File: src/core_controller.cpp
#include "core_controller.h"
#include <xmlrpcpp/XmlRpcValue.h>
using namespace std;

CoreControllerNode::CoreControllerNode()
{
    // 读取yaml文件中事先设定的参数
    initFromYaml();

    // 初始化各PoseStamped消息的header(用于rviz可视化)
    uav_target_pose_global_.header.seq = 1;
    uav_target_pose_global_.header.frame_id = "map";
    uav_target_pose_local_.header.seq = 1;
    uav_target_pose_local_.header.frame_id = "map";
    current_pose_.header.seq = 1;
    current_pose_.header.frame_id = "map";

    is_traj_local_init_ = false;
    target_state_ = 0;
    min_dist_ = 0.3;
    min_yaw_diff_ = 5;
    target_dist_threshold_ = 1;
    flight_state_ = WAITING;
    last_request_ = ros::Time::now();

    pub_interval_ = ros::Duration(0.02);
    decision_interval_ = ros::Duration(0.05);

    uav_target_pose_global_pub_ = n_.advertise<geometry_msgs::PoseStamped>("/core_controller/uav_traj_pose", 100);
    uav_target_pose_local_pub_ = n_.advertise<geometry_msgs::PoseStamped>("/mavros/setpoint_position/local", 100);
    gazeboposePub_ = n_.advertise<geometry_msgs::PoseStamped>("/core_controller/uav_pose_gazebo", 100);

    mavstateSub_ = n_.subscribe("/mavros/state", 1, &CoreControllerNode::mavstateCallback, this, ros::TransportHints().tcpNoDelay());  
    gazeboposeSub_ = n_.subscribe("/gazebo/model_states", 1, &CoreControllerNode::gazeboposeCallback, this, ros::TransportHints().tcpNoDelay());  
    targetposeSub_ = n_.subscribe("/target_location/camera", 1, &CoreControllerNode::targetposeCallback, this, ros::TransportHints().tcpNoDelay());

    arming_client_ = n_.serviceClient<mavros_msgs::CommandBool>("/mavros/cmd/arming");
    land_client_ = n_.serviceClient<mavros_msgs::CommandBool>("/mavros/cmd/land");
    set_mode_client_ = n_.serviceClient<mavros_msgs::SetMode>("/mavros/set_mode");

    statusloop_timer_ = n_.createTimer(ros::Duration(1), &CoreControllerNode::statusloopCallback, this);
    publishloop_timer_ = n_.createTimer(pub_interval_, &CoreControllerNode::publishloopCallback, this);
    decisionloop_timer_ = n_.createTimer(decision_interval_, &CoreControllerNode::decisionloopCallback, this);

    ros::Rate rate(20);
    for(int i = 0; ros::ok() && i < 10; i++)
    {
        uav_target_pose_local_pub_.publish(uav_target_pose_local_);
        ros::spinOnce();
        rate.sleep();
    }
}

void CoreControllerNode::initFromYaml()
{
    // 初始化世界坐标系下的路径traj_global_和yaw_global_
    XmlRpc::XmlRpcValue traj_global_temp;
    Eigen::Vector3d v;
    if(!n_.getParam("/core_controller/trajectory", traj_global_temp))
        ROS_ERROR("Failed to get param.");
    ROS_ASSERT(traj_global_temp.getType() == XmlRpc::XmlRpcValue::TypeArray);
    ROS_ASSERT(traj_global_temp[0].getType() == XmlRpc::XmlRpcValue::TypeArray);
    ROS_ASSERT(traj_global_temp[0][0].getType() == XmlRpc::XmlRpcValue::TypeDouble || 
        traj_global_temp[0][0].getType() == XmlRpc::XmlRpcValue::TypeInt);

    for(int i = 0; i < traj_global_temp.size(); i++)
    {
        v << double(traj_global_temp[i][0]), double(traj_global_temp[i][1]), double(traj_global_temp[i][2]);
        traj_global_.push_back(v);
        yaw_global_.push_back(double(traj_global_temp[i][3]));
    }
    // 初始化相机位姿
    XmlRpc::XmlRpcValue pose_temp;
    if(!n_.getParam("/core_controller/camera_pose", pose_temp))
        ROS_ERROR("Failed to get param.");
    ROS_ASSERT(pose_temp.getType() == XmlRpc::XmlRpcValue::TypeArray);
    ROS_ASSERT(pose_temp[0].getType() == XmlRpc::XmlRpcValue::TypeDouble || 
        pose_temp[0].getType() == XmlRpc::XmlRpcValue::TypeInt);
    t_uc_ << double(pose_temp[0]), double(pose_temp[1]), double(pose_temp[2]);
    R_uc_ = Eigen::AngleAxisd(double(pose_temp[3]), Eigen::Vector3d::UnitZ()) *
           Eigen::AngleAxisd(double(pose_temp[4]), Eigen::Vector3d::UnitY()) *
           Eigen::AngleAxisd(double(pose_temp[5]), Eigen::Vector3d::UnitX());  // 欧拉角转旋转矩阵
}

double CoreControllerNode::targetDist()
{
    return sqrt((current_pose_.pose.position.x - uav_target_pose_global_.pose.position.x)
        * (current_pose_.pose.position.x - uav_target_pose_global_.pose.position.x)
        + (current_pose_.pose.position.y - uav_target_pose_global_.pose.position.y)
        * (current_pose_.pose.position.y - uav_target_pose_global_.pose.position.y)
        + (current_pose_.pose.position.z - uav_target_pose_global_.pose.position.z)
        * (current_pose_.pose.position.z - uav_target_pose_global_.pose.position.z));
}

double CoreControllerNode::targetYawDiff()
{
    double uav_target_y, current_y;
    uav_target_y = tf::getYaw(uav_target_pose_global_.pose.orientation);
    current_y = tf::getYaw(current_pose_.pose.orientation);
    double diff = (uav_target_y - current_y) / M_PI * 180.0;
    if(diff > 180.0)
        return diff - 360.0;
    else if(diff < -180.0)
        return diff + 360.0;
    else
        return diff;
}

bool CoreControllerNode::isAlreadyDetected(Eigen::Vector3d pos)
{
    for(auto iter = already_detected_target_.begin(); iter != already_detected_target_.end(); iter++)
    {
        if((pos - *iter).norm() < target_dist_threshold_)
            return true;
    }
    return false;
}

void CoreControllerNode::popUAVTargetPose()
{
    if(traj_global_.empty() || traj_local_.empty())
    {
        return;
    }
    
    Eigen::Vector3d v;

    v = traj_global_.front();
    uav_target_pose_global_.header.stamp = ros::Time::now();
    uav_target_pose_global_.pose.position.x = v[0];
    uav_target_pose_global_.pose.position.y = v[1];
    uav_target_pose_global_.pose.position.z = v[2];
    uav_target_pose_global_.pose.orientation = tf::createQuaternionMsgFromYaw((yaw_global_.front()/180.0)*M_PI);
    traj_global_.pop_front();
    yaw_global_.pop_front();

    v = traj_local_.front();
    uav_target_pose_local_.header.stamp = ros::Time::now();
    uav_target_pose_local_.pose.position.x = v[0];
    uav_target_pose_local_.pose.position.y = v[1];
    uav_target_pose_local_.pose.position.z = v[2];
    uav_target_pose_local_.pose.orientation = tf::createQuaternionMsgFromYaw((yaw_local_.front()/180.0)*M_PI);
    traj_local_.pop_front();
    yaw_local_.pop_front();
}

void CoreControllerNode::globalTrajPushFront(Eigen::Vector3d pos, double yaw)
{
    // global
    traj_global_.push_front(pos);
    yaw_global_.push_front(yaw);
    // local
    if(is_traj_local_init_)
    {
        traj_local_.push_front(R_init_.inverse()*(pos-t_init_));
        yaw_local_.push_front(-yaw_init_+yaw < 0 ? -yaw_init_+yaw+360 : -yaw_init_+yaw);
    }
}

void CoreControllerNode::localTrajPushFront(Eigen::Vector3d pos, double yaw)
{
    // global
    traj_global_.push_front(R_init_*pos+t_init_);
    yaw_global_.push_front(yaw_init_+yaw > 360 ? yaw_init_+yaw-360 : yaw_init_+yaw);
    // local
    if(is_traj_local_init_)
    {
        traj_local_.push_front(pos);
        yaw_local_.push_front(yaw);
    }
}

void CoreControllerNode::globalTrajPushBack(Eigen::Vector3d pos, double yaw)
{
    // global
    traj_global_.push_back(pos);
    yaw_global_.push_back(yaw);
    // local
    if(is_traj_local_init_)
    {
        traj_local_.push_back(R_init_.inverse()*(pos-t_init_));
        yaw_local_.push_back(-yaw_init_+yaw < 0 ? -yaw_init_+yaw+360 : -yaw_init_+yaw);
    }
}

void CoreControllerNode::localTrajPushBack(Eigen::Vector3d pos, double yaw)
{
    // global
    traj_global_.push_back(R_init_*pos+t_init_);
    yaw_global_.push_back(yaw_init_+yaw > 360 ? yaw_init_+yaw-360 : yaw_init_+yaw);
    // local
    if(is_traj_local_init_)
    {
        traj_local_.push_back(pos);
        yaw_local_.push_back(yaw);
    }
}

void CoreControllerNode::clearTraj()
{
    traj_global_.clear();
    yaw_global_.clear();
    traj_local_.clear();
    yaw_local_.clear();
}

void CoreControllerNode::mavstateCallback(const mavros_msgs::State::ConstPtr& msg)
{
    current_state_ = *msg;
}

void CoreControllerNode::gazeboposeCallback(const gazebo_msgs::ModelStates::ConstPtr& msg)
{
    int i;
    for(i = 0; i < msg->name.size(); i++)
    {
        if(msg->name[i] == "iris")
            break;
    }
    if(i < msg->name.size())
    {
        geometry_msgs::Pose pose = msg->pose[i];
        current_pose_.header.stamp = ros::Time::now();
        current_pose_.pose = pose;
        t_wu_ << pose.position.x, pose.position.y, pose.position.z;
        Eigen::Quaterniond q(pose.orientation.w, pose.orientation.x, pose.orientation.y, pose.orientation.z);
        R_wu_ = q.toRotationMatrix();

        if(!is_traj_local_init_) 
        {
            // 利用第一次收到的世界坐标将traj_global_、yaw_global_中的轨迹信息转化为本地坐标（以无人机初始化位置为原点的坐标系）
            t_init_ = t_wu_;
            yaw_init_ = tf::getYaw(pose.orientation)/M_PI*180.0;
            R_init_ = Eigen::AngleAxisd(yaw_init_/180.0*M_PI, Eigen::Vector3d::UnitZ()).toRotationMatrix();
            for(auto trajPtr = traj_global_.begin(); trajPtr != traj_global_.end(); trajPtr++)
            {
                traj_local_.push_back(R_init_.inverse()*(*trajPtr-t_init_));
            }
            for(auto yawPtr = yaw_global_.begin(); yawPtr != yaw_global_.end(); yawPtr++)
            {
                yaw_local_.push_back(-yaw_init_+*yawPtr < 0 ? -yaw_init_+*yawPtr+360 : -yaw_init_+*yawPtr);
            }
            // 初始化第一个飞行目标点
            popUAVTargetPose();
        }
        is_traj_local_init_ = true;
        
    }
}

void CoreControllerNode::targetposeCallback(const uav_sim::PointWithState::ConstPtr& msg)
{
    target_state_ = msg->state;
    if(target_state_)
    {
        Eigen::Vector3d p_ct(msg->position.x, msg->position.y, msg->position.z), p_wt;
        p_wt = R_wu_*(R_uc_*p_ct+t_uc_)+t_wu_;
        // p_wt = R_wu_*p_ct+t_wu_;
        if(!this->isAlreadyDetected(p_wt))
        {
            already_detected_target_.push_back(p_wt);
            ROS_INFO("Target detected at %lf, %lf, %lf", p_wt[0], p_wt[1], p_wt[2]);
        }
    }
}

void CoreControllerNode::statusloopCallback(const ros::TimerEvent& event)
{
    switch(flight_state_)
    {
    case WAITING:
        arm_cmd_.request.value = true;
        offb_set_mode_.request.custom_mode = "OFFBOARD";
        if( current_state_.mode != "OFFBOARD" && (ros::Time::now() - last_request_ > ros::Duration(4.0)) && is_traj_local_init_)
        {
    	    if( set_mode_client_.call(offb_set_mode_) && offb_set_mode_.response.mode_sent)
            {
    	        ROS_INFO("Offboard enabled");
	    }
	    last_request_ = ros::Time::now();
        } 
        else 
        {
	    if( !current_state_.armed && (ros::Time::now() - last_request_ > ros::Duration(4.0)))
	    {
	        if( arming_client_.call(arm_cmd_) && arm_cmd_.response.success)
	        {
	            ROS_INFO("Vehicle armed");
                    flight_state_ = FLYING;
	        }
	        last_request_ = ros::Time::now();
	    }    
        }
        break;
    case FLYING:
        break;
    case READY_TO_LAND:
        ROS_INFO("Ready to land");
        offb_set_mode_.request.custom_mode = "AUTO.LAND";   
        if( current_state_.mode != "AUTO.LAND")
        {
            if(set_mode_client_.call(offb_set_mode_) && offb_set_mode_.response.mode_sent)
	    {
	        ROS_INFO("Vehicle landing");
                flight_state_ = LANDING;
	    }
	    last_request_ = ros::Time::now();
        }
        break;
    case LANDING:
        break;
    default:
        break;
    }
}

void CoreControllerNode::publishloopCallback(const ros::TimerEvent& event)
{
    if(is_traj_local_init_)
    {
        uav_target_pose_global_pub_.publish(uav_target_pose_global_);
        uav_target_pose_local_pub_.publish(uav_target_pose_local_);
        gazeboposePub_.publish(current_pose_);
    }
}

void CoreControllerNode::decisionloopCallback(const ros::TimerEvent& event)
{
    // 如果当前无人机位姿与下一路径位姿差别足够小，做出决策
    if(flight_state_ == FLYING && targetDist() < min_dist_ && fabs(targetYawDiff()) < min_yaw_diff_) 
    {
        if(traj_local_.empty())  // 已经到达最终路径点，准备降落
        {
            flight_state_ = READY_TO_LAND;
        }
        else // 从队列中取出下一个路径位姿
        {
            popUAVTargetPose();
        }
    }
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "core_controller_node");

    CoreControllerNode f;

    if(!ros::ok())
    {
	return 0;
    }
    ROS_INFO("flight controller node set up already.");
    ros::spin();

    return 0;
}

File: scripts/core_controller.py
#!/usr/bin/python
#-*- encoding: utf8 -*-

from scipy.spatial.transform import Rotation as R
import rospy
import numpy as np
import math
import collections
from enum import Enum
from geometry_msgs.msg import PoseStamped
from mavros_msgs.msg import State
from gazebo_msgs.msg import ModelStates
from uav_sim.msg import PointWithState
from mavros_msgs.srv import CommandBool
from mavros_msgs.srv import SetMode

class CoreControllerNode:

    class FlightState(Enum):
        WAITING = 1
        FLYING = 2
        READY_TO_LAND = 3
        LANDING = 4


    def __init__(self):
        rospy.init_node('core_controller_node', anonymous=True)

        #以下4个变量为无人机路径位姿队列，需要同时进行维护
        self.traj_global_ = collections.deque()
        self.yaw_global_ = collections.deque()  #角度制
        self.traj_local_ = collections.deque()
        self.yaw_local_ = collections.deque()  #角度制

        #读取yaml文件中事先设定的参数
        self.initFromYaml();

        #初始化各PoseStamped消息及其header(用于rviz可视化)
        self.uav_target_pose_global_ = PoseStamped()
        self.uav_target_pose_global_.header.seq = 1
        self.uav_target_pose_global_.header.frame_id = 'map'
        self.uav_target_pose_global_.pose.orientation.w = 1
        self.uav_target_pose_local_ = PoseStamped()
        self.uav_target_pose_local_.header.seq = 1
        self.uav_target_pose_local_.header.frame_id = 'map'
        self.uav_target_pose_local_.pose.orientation.w = 1
        self.current_pose_ = PoseStamped()  #世界坐标
        self.current_pose_.header.seq = 1
        self.current_pose_.header.frame_id = 'map'
        self.current_pose_.pose.orientation.w = 1

        #目标检测相关变量
        self.already_detected_target_ = []  #已侦测到的目标坐标
        self.target_state_ = 0  #目标状态（1=侦测到，0=未侦测到）
        self.min_dist_ = 0.3  #在无人机与当前目标点（uav_target_pose_）距离小于min_dist_且yaw差距小于min_yaw_diff_时，当前飞行目标点改为路径队列中的下一个点
        self.min_yaw_diff_ = 5
        self.target_dist_threshold_ = 1

        #无人机在世界坐标系下的位姿
        self.R_wu_ = R.from_quat([0, 0, 0, 1])
        self.t_wu_ = np.zeros([3], dtype = 'float')

        #无人机的初始位姿，在第一次收到相关topic后初始化，用于进行Gazebo世界坐标与MAVROS本地坐标的转换
        self.t_init_ = np.zeros([3], dtype = 'float')
        self.yaw_init_ = 0.0  #角度制
        self.R_init_ = R.from_quat([0, 0, 0, 1])
        self.is_traj_local_init_ = False

        #用于发送起降命令和追踪飞行状态
        self.last_request_ = rospy.Time.now()
        self.offb_set_mode_ = SetMode()
        self.arm_cmd_ = CommandBool()
        self.current_state_ = State()
        self.flight_state_ = self.FlightState.WAITING

        #其他ROS相关内容
        self.pub_interval_ = rospy.Duration(0.02)
        self.decision_interval_ = rospy.Duration(0.05)

        self.uav_target_pose_global_pub_ = rospy.Publisher('/core_controller/uav_traj_pose', PoseStamped, queue_size=100)
        self.uav_target_pose_local_pub_ = rospy.Publisher('/mavros/setpoint_position/local', PoseStamped, queue_size=100)
        self.gazeboposePub_ = rospy.Publisher('/core_controller/uav_pose_gazebo', PoseStamped, queue_size=100)

        self.mavstateSub_ = rospy.Subscriber('/mavros/state', State, self.mavstateCallback)
        self.gazeboposeSub_ = rospy.Subscriber('/gazebo/model_states', ModelStates, self.gazeboposeCallback)
        self.targetposeSub_ = rospy.Subscriber('/target_location/location', PointWithState, self.targetposeCallback)

        self.arming_client_ = rospy.ServiceProxy('/mavros/cmd/arming', CommandBool)
        self.land_client_ = rospy.ServiceProxy('/mavros/cmd/land', CommandBool)
        self.set_mode_client_ = rospy.ServiceProxy('/mavros/set_mode', SetMode)

        self.statusloop_timer_ = rospy.Timer(rospy.Duration(1), self.statusloopCallback)
        self.publishloop_timer_ = rospy.Timer(self.pub_interval_, self.publishloopCallback)
        self.decisionloop_timer_ = rospy.Timer(self.decision_interval_, self.decisionloopCallback)

        rate = rospy.Rate(20);
        for i in range(10):
            self.uav_target_pose_local_pub_.publish(self.uav_target_pose_local_);
            rate.sleep()

        try:
            rospy.spin()
        except KeyboardInterrupt:
            pass


    #从yaml文件中初始化部分参数
    def initFromYaml(self):
        #初始化世界坐标系下的路径traj_global_和yaw_global_
        traj_global_temp = rospy.get_param('/core_controller/trajectory')
        for pt in traj_global_temp:
            self.traj_global_.append(np.array(pt[:3], dtype = 'float'))
            self.yaw_global_.append(float(pt[3]))
        
        #初始化相机位姿
        pose_temp = rospy.get_param('/core_controller/camera_pose')
        self.t_uc_ = np.array(pose_temp[:3], dtype = 'float')
        self.R_uc_ = R.from_euler('zyx', pose_temp[3:6], degrees = True)
        pass


    def targetDist(self):
        return math.sqrt((self.current_pose_.pose.position.x - self.uav_target_pose_global_.pose.position.x) ** 2 \
                       + (self.current_pose_.pose.position.y - self.uav_target_pose_global_.pose.position.y) ** 2 \
                       + (self.current_pose_.pose.position.z - self.uav_target_pose_global_.pose.position.z) ** 2)


    def targetYawDiff(self):
        #四元数转欧拉角
        uav_target_r = R.from_quat([self.uav_target_pose_global_.pose.orientation.x, \
                                   self.uav_target_pose_global_.pose.orientation.y, \
                                   self.uav_target_pose_global_.pose.orientation.z, \
                                   self.uav_target_pose_global_.pose.orientation.w])
        current_r = R.from_quat([self.current_pose_.pose.orientation.x, \
                                self.current_pose_.pose.orientation.y, \
                                self.current_pose_.pose.orientation.z, \
                                self.current_pose_.pose.orientation.w])
        [uav_target_y, pitch, roll] = uav_target_r.as_euler('zyx', degrees = True)
        [current_y, pitch, roll] = current_r.as_euler('zyx', degrees = True)
        diff = uav_target_y - current_y
        if diff > 180.0:
            return diff - 360.0
        elif diff < -180.0:
            return diff + 360.0
        else:
            return diff


    def isAlreadyDetected(self, pos):
        for target in self.already_detected_target_:
            if np.linalg.norm(pos - target) < self.target_dist_threshold_:
                return True
        return False


    #从无人机路径位姿队列头部取出一个点作为下一目标点，更新uav_target_pose_global_和uav_target_pose_local_
    def popUAVTargetPose(self):
        if len(self.traj_global_) == 0 or len(self.traj_local_) == 0:
            return
        
        v = self.traj_global_.popleft()
        yaw = self.yaw_global_.popleft()
        (x, y, z, w) = (R.from_euler('zyx', [yaw, 0, 0], degrees = True)).as_quat()
        self.uav_target_pose_global_.header.stamp = rospy.Time.now();
        self.uav_target_pose_global_.pose.position.x = v[0]
        self.uav_target_pose_global_.pose.position.y = v[1]
        self.uav_target_pose_global_.pose.position.z = v[2]
        self.uav_target_pose_global_.pose.orientation.x = x
        self.uav_target_pose_global_.pose.orientation.y = y
        self.uav_target_pose_global_.pose.orientation.z = z
        self.uav_target_pose_global_.pose.orientation.w = w
        
        v = self.traj_local_.popleft()
        yaw = self.yaw_local_.popleft()
        (x, y, z, w) = (R.from_euler('zyx', [yaw, 0, 0], degrees = True)).as_quat()
        self.uav_target_pose_local_.header.stamp = rospy.Time.now();
        self.uav_target_pose_local_.pose.position.x = v[0]
        self.uav_target_pose_local_.pose.position.y = v[1]
        self.uav_target_pose_local_.pose.position.z = v[2]
        self.uav_target_pose_local_.pose.orientation.x = x
        self.uav_target_pose_local_.pose.orientation.y = y
        self.uav_target_pose_local_.pose.orientation.z = z
        self.uav_target_pose_local_.pose.orientation.w = w
        

    def globalTrajPushFront(self, pos, yaw):
        #global
        self.traj_global_.appendleft(pos);
        self.yaw_global_.appendleft(yaw);
        #local
        if is_traj_local_init_:
            self.traj_local_.appendleft(np.matmul(self.R_init_.inv().as_dcm(), pos-self.t_init_))
            self.yaw_local_.appendleft(-self.yaw_init_+yaw+360 if (-self.yaw_init_+yaw < 0) else -self.yaw_init_+yaw)


    def localTrajPushFront(self, pos, yaw):
        #global
        self.traj_global_.appendleft(np.matmul(self.R_init_.as_dcm(), pos) + self.t_init_);
        self.yaw_global_.appendleft(self.yaw_init_+yaw-360 if (self.yaw_init_+yaw > 360) else self.yaw_init_+yaw);
        #local
        if is_traj_local_init_:
            self.traj_local_.appendleft(pos)
            self.yaw_local_.appendleft(yaw)


    def globalTrajPushBack(self, pos, yaw):
        #global
        self.traj_global_.append(pos);
        self.yaw_global_.append(yaw);
        #local
        if is_traj_local_init_:
            self.traj_local_.append(np.matmul(self.R_init_.inv().as_dcm(), pos-self.t_init_))
            self.yaw_local_.append(-self.yaw_init_+yaw+360 if (-self.yaw_init_+yaw < 0) else -self.yaw_init_+yaw)


    def localTrajPushBack(self, pos, yaw):
        #global
        self.traj_global_.append(np.matmul(self.R_init_.as_dcm(), pos) + self.t_init_);
        self.yaw_global_.append(self.yaw_init_+yaw-360 if (self.yaw_init_+yaw > 360) else self.yaw_init_+yaw);
        #local
        if is_traj_local_init_:
            self.traj_local_.append(pos)
            self.yaw_local_.append(yaw)


    #清空无人机路径位姿队列
    def clearTraj(self):
        self.traj_global_ = collections.deque()
        self.yaw_global_ = collections.deque()
        self.traj_local_ = collections.deque()
        self.yaw_local_ = collections.deque()


    def mavstateCallback(self, msg):
        self.current_state_ = msg


    def gazeboposeCallback(self, msg):
        if 'iris' in msg.name:
            i = msg.name.index('iris')
            pose = msg.pose[i]
            self.current_pose_.header.stamp = rospy.Time.now()
            self.current_pose_.pose = pose
            self.t_wu_ = np.array([pose.position.x, pose.position.y, pose.position.z])
            self.R_wu_ = R.from_quat([pose.orientation.x, pose.orientation.y, pose.orientation.z, pose.orientation.w])
            if not self.is_traj_local_init_:
                #利用第一次收到的世界坐标将traj_global_、yaw_global_中的轨迹信息转化为本地坐标（以无人机初始化位置为原点的坐标系）
                self.t_init_ = self.t_wu_
                (self.yaw_init_, pitch, roll) = self.R_wu_.as_euler('zyx', degrees = True)
                self.R_init_ = self.R_wu_
                for trajPtr in self.traj_global_:
                    self.traj_local_.append(np.matmul(self.R_init_.inv().as_dcm(), trajPtr-self.t_init_))
                for yawPtr in self.yaw_global_:
                    self.yaw_local_.append(-self.yaw_init_+yawPtr+360 if (-self.yaw_init_+yawPtr < 0) else -self.yaw_init_+yawPtr)
                #初始化第一个飞行目标点
                self.popUAVTargetPose()
            self.is_traj_local_init_ = True


    def targetposeCallback(self, msg):
        self.target_state_ = msg.state
        if self.target_state_:
            p_ct = np.array([msg.position.x, msg.position.y, msg.position.z])
            p_ut = np.matmul(self.R_uc_.as_dcm(), p_ct) + self.t_uc_
            p_wt = np.matmul(self.R_wu_.as_dcm(), p_ut) + self.t_wu_
            if not self.isAlreadyDetected(p_wt):
                self.already_detected_target_.append(p_wt)
                log_str = 'Target detected at %f, %f, %f' % (p_wt[0], p_wt[1], p_wt[2])
                rospy.loginfo(log_str)


    #定时检测无人机状态，根据情况发布arming和起降命令
    def statusloopCallback(self, event):
        if self.flight_state_ == self.FlightState.WAITING:
            if self.current_state_.mode != 'OFFBOARD' and (rospy.Time.now() - self.last_request_ > rospy.Duration(4.0)) and self.is_traj_local_init_:
                response = self.set_mode_client_(0, 'OFFBOARD')  #请求解锁
                if response.mode_sent:
                    rospy.loginfo('Offboard enabled')
                self.last_request_ = rospy.Time.now()
            else:
                if (not self.current_state_.armed) and rospy.Time.now() - self.last_request_ > rospy.Duration(4.0):
                    response = self.arming_client_(True)  #请求起飞
                    if response.success:
                        rospy.loginfo('Vehicle armed')
                        self.flight_state_ = self.FlightState.FLYING
                    self.last_request_ = rospy.Time.now()
        elif self.flight_state_ == self.FlightState.FLYING:
            pass
        elif self.flight_state_ == self.FlightState.READY_TO_LAND:
            rospy.loginfo('Ready to land')
            if self.current_state_.mode != 'AUTO.LAND':
                response = self.set_mode_client_(0, 'AUTO.LAND')  #请求降落
                if response.mode_sent:
                    rospy.loginfo('Vehicle landing')
                    self.flight_state_ = self.FlightState.LANDING
                self.last_request_ = rospy.Time.now()
        else:
            pass
            

    def publishloopCallback(self, event):
        if self.is_traj_local_init_:
            self.uav_target_pose_global_pub_.publish(self.uav_target_pose_global_);
            self.uav_target_pose_local_pub_.publish(self.uav_target_pose_local_);
            self.gazeboposePub_.publish(self.current_pose_);


    #决策策略可以在此函数中实现
    def decisionloopCallback(self, event):
        #如果当前无人机位姿与下一路径位姿差别足够小，做出决策
        if self.flight_state_ == self.FlightState.FLYING and self.targetDist() < self.min_dist_ \
                   and abs(self.targetYawDiff()) < self.min_yaw_diff_: 
            if len(self.traj_local_) == 0:  #已经到达最终路径点，准备降落
                self.flight_state_ = self.FlightState.READY_TO_LAND;
            else: #从队列中取出下一个路径位姿
                self.popUAVTargetPose();


if __name__ == '__main__':
    cc = CoreControllerNode()

